realm/src/main/java/io/realm/BaseRealm.java

```diff
@@ -436,19 +436,21 @@ protected synchronized void acquireFileReference(RealmConfiguration configuratio
      * Releases a reference to the Realm file. If reference count reaches 0 any cached configurations
      * will be removed.
      */
-    protected synchronized void releaseFileReference() {
-        String canonicalPath = configuration.getPath();
-        List<RealmConfiguration> pathConfigurationCache = globalPathConfigurationCache.get(canonicalPath);
-        pathConfigurationCache.remove(configuration);
-        if (pathConfigurationCache.isEmpty()) {
-            globalPathConfigurationCache.remove(canonicalPath);
-        }
+    protected void releaseFileReference() {
+        synchronized (BaseRealm.class) {
+            String canonicalPath = configuration.getPath();
+            List<RealmConfiguration> pathConfigurationCache = globalPathConfigurationCache.get(canonicalPath);
+            pathConfigurationCache.remove(configuration);
+            if (pathConfigurationCache.isEmpty()) {
+                globalPathConfigurationCache.remove(canonicalPath);
+            }
 
-        Integer refCount = globalRealmFileReferenceCounter.get(canonicalPath);
-        if (refCount == null || refCount == 0) {
-            throw new IllegalStateException("Trying to release a Realm file that is already closed");
+            Integer refCount = globalRealmFileReferenceCounter.get(canonicalPath);
```

Comment:
Actually `acquireFileReference` is only called from `createAndValidate` which has a lock on `BaseRealm.class` so right now it is safe. 

I could see why we want to have it inside this method as well though.

In general I think we should take a deep hard look at how we do locking currently. We have a mix of Concurrent* and locks that are very hard to reason about. So while it might be correct now, the chance of future bugs are very high, and it is impossible to write unit tests for.

I think it would probably to make the locking more coarse-grained vs. the ability to actually reason about it, but that should be another issue.